apps/cmp.c

```diff
@@ -0,0 +1,3256 @@
+    }
+    if (opt_srv_cert != NULL) {
+        X509 *srv_cert = load_cert_pwd(opt_srv_cert, opt_srv_keypass,
+                                       "certificate of the server");
+        if (srv_cert == NULL || !OSSL_CMP_CTX_set1_cert(ctx, srv_cert)) {
+            X509_free(srv_cert);
+            goto err;
+        }
+        X509_free(srv_cert);
+    }
+    if (opt_srv_key != NULL) {
+        EVP_PKEY *pkey = load_key_pwd(opt_srv_key, opt_keyform,
+                                      opt_srv_keypass,
+                                      e, "private key for server cert");
+
+        if (pkey == NULL || !OSSL_CMP_CTX_set1_pkey(ctx, pkey)) {
+            EVP_PKEY_free(pkey);
+            goto err;
+        }
+        EVP_PKEY_free(pkey);
+    }
+    cleanse(opt_srv_keypass);
+
+    if (opt_srv_trusted != NULL) {
+        X509_STORE *ts =
+            load_certstore(opt_srv_trusted, "certificates trusted by server");
+
+        if (ts == NULL)
+            goto err;
+        if (!set1_store_parameters(ts)
+                || !truststore_set_host_etc(ts, NULL)
+                || !OSSL_CMP_CTX_set0_trustedStore(ctx, ts)) {
+            X509_STORE_free(ts);
+            goto err;
+        }
+    } else {
+        CMP_warn("server will not be able to handle signature-protected requests since -srv_trusted is not given");
+    }
+    if (!setup_certs(opt_srv_untrusted, "untrusted certificates", ctx,
+                     (add_X509_stack_fn_t)OSSL_CMP_CTX_set1_untrusted_certs,
+                     NULL))
+        goto err;
+
+    if (opt_rsp_cert == NULL) {
+        CMP_err("must give -rsp_cert for mock server");
+        goto err;
+    } else {
+        X509 *cert = load_cert_pwd(opt_rsp_cert, opt_keypass,
+                                   "cert to be returned by the mock server");
+
+        if (cert == NULL)
+            goto err;
+        /* from server perspective the server is the client */
+        if (!ossl_cmp_mock_srv_set1_certOut(srv_ctx, cert)) {
+            X509_free(cert);
+            goto err;
+        }
+        X509_free(cert);
+    }
+    /* TODO find a cleaner solution not requiring type casts */
+    if (!setup_certs(opt_rsp_extracerts,
+                     "CMP extra certificates for mock server", srv_ctx,
+                     (add_X509_stack_fn_t)ossl_cmp_mock_srv_set1_chainOut,
+                     NULL))
+        goto err;
+    if (!setup_certs(opt_rsp_capubs, "caPubs for mock server", srv_ctx,
+                     (add_X509_stack_fn_t)ossl_cmp_mock_srv_set1_caPubsOut,
+                     NULL))
+        goto err;
+    (void)ossl_cmp_mock_srv_set_pollCount(srv_ctx, opt_poll_count);
+    (void)ossl_cmp_mock_srv_set_checkAfterTime(srv_ctx, opt_check_after);
+    if (opt_grant_implicitconf)
+        (void)OSSL_CMP_SRV_CTX_set_grant_implicit_confirm(srv_ctx, 1);
+
+    if (opt_failure < -1 || opt_failure > OSSL_CMP_PKIFAILUREINFO_MAX) {
+        CMP_err1("-failure out of range, should be >= 0 and <= %d",
```

Comment:
test is not consistent with error message, when `opt_failure`  is -1

Reply:
Yes, but this was on purpose since -1 was used to indicate the option was not explicitly set.
I've changed this to use INT_MIN for that purpose and also re-arranged the condition checks and added a comment, which should make things more clear.